src/celeritas/ext/RootUniquePtr.hh

```diff
@@ -0,0 +1,70 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022-2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/RootUniquePtr.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+
+#include "celeritas_config.h"
+#include "corecel/Assert.hh"
+
+// Forward-declare ROOT
+class TObject;
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+//! Helpers to prevent ROOT from propagating to downstream code.
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Call `TObject->Write()` before deletion. Used by TFile and TTree writer
+ * classes.
+ */
+template<class T>
+struct WriteRootDeleter
+{
+    void operator()(T* ptr);
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Deleter only. Used by TFile and TTree reader classes or TObjects that should
+ * not invoke `Write()`.
+ */
+template<class T>
+struct ReadRootDeleter
+{
+    void operator()(T* ptr);
+};
+
+//---------------------------------------------------------------------------//
+// Type aliases
+template<class T>
+using UPRootWriter = std::unique_ptr<T, WriteRootDeleter<T>>;
```

Comment:
Either
```suggestion
using UPRootWriter = std::unique_ptr<T, WriteRootDeleter>; //  Using TObject* as the parameter.
```
or
```suggestion
using UPRootWriter = std::unique_ptr<T, WriteBeforeDelete<T>>;
```
as it technical can apply to any type that has a 'Write' function.

Reply:
We can't do the first option because knowing that `TFile*` can be implicitly cast to `TObject*` requires the definition of `TFile`, which is what we're hoping to avoid including.

Reply:
Then why bother putting `TObject` in the operator implementation (see comment https://github.com/celeritas-project/celeritas/pull/610#discussion_r1088429497).   I.e.
```
template <typename T> struct WriteThenDelete
 {
     void operator()(T* ptr);
 };
 ```

Reply:
Yeah I think my comment there was in error: you're correct.